src/inet/InetInterface.cpp

```diff
@@ -105,15 +109,41 @@ bool InterfaceIterator::Next()
     // TODO : Cleanup #17346
     return false;
 }
+
+InterfaceAddressIterator::InterfaceAddressIterator()
+{
+    mNetifAddrList = nullptr;
+    mCurAddr       = nullptr;
+}
+
 bool InterfaceAddressIterator::HasCurrent()
 {
-    return mIntfIter.HasCurrent();
+    return (mNetifAddrList != nullptr) ? (mCurAddr != nullptr) : Next();
 }
 
 bool InterfaceAddressIterator::Next()
 {
-    // TODO : Cleanup #17346
-    return false;
+    if (mNetifAddrList == nullptr)
+    {
+        mNetifAddrList = otIp6GetUnicastAddresses(Inet::globalOtInstance);
+
+        if (mNetifAddrList == nullptr)
+        {
+            return false;
+        }
+        mCurAddr = const_cast<otNetifAddress *>(mNetifAddrList);
```

Comment:
why do we need to const-cast? Thinking removing constness is a code smell. If we do need it, a comment is required to explain.

Reply:
Could mabye mCurAddr be const?

Reply:
cast removed